storage/repair.go

```diff
@@ -124,23 +128,266 @@ func (r shardRepairer) Repair(
+	initialCap := diffRes.NumBlocks
+	reqBlocks := r.newAdminRepairBlocksRequest(pendingBlockReplicas, hashToIDMap, initialCap)
+	repairSummary.NumRequested = int64(len(reqBlocks))
+
+	// stream over the requested replicas
+	blocksIter, err := session.FetchRepairBlocksFromPeers(namespace, shard.ID(), reqBlocks, r.rpopts.ResultOptions())
+	if err != nil {
+		return repairSummary, err
+	}
+
+	for blocksIter.Next() {
+		host, id, blk := blocksIter.Current()
+		// TODO(prateek): does Close() need to be called on blk
+		// i.e. figure out ownership - is it reset by iterator or not
+
+		blkID := blockID{id.Hash(), blk.StartTime()}
+		// ensure the replica was requested, should never happen
+		if hs, ok := pendingBlockReplicas[blkID]; !ok || !hs.contains(host) {
+			logger.WithFields(
+				xlog.NewLogField("id", id.String()),
+				xlog.NewLogField("host", host.String()),
+				xlog.NewLogField("blockID", blkID),
+			).Warnf("received un-requested block, session.FetchRepairBlockFromPeers violated contract, skipping.")
+			repairSummary.NumUnrequested++
+			continue
+		}
+
+		// mark the block replica as received
+		pendingBlockReplicas[blkID].remove(host)
+
+		// TODO(prateek): current implementation of UpdateSeries marks shards flushed internally,
+		// this is something we should amortize to be done at the end of iterating through this function.
+		// concern being, we want to minimize the number of flushes we can potentially induce
+		markFlushStateDirty := true
+		if err := shard.UpdateSeries(id, blk, markFlushStateDirty); err != nil {
+			repErr := fmt.Errorf(
+				"unable to update series [ id = %v, block_start = %v, err = %v ]",
+				id.String(), blkID.start, err)
+			multiErr = multiErr.Add(repErr)
+			logger.Warnf("%v", repErr)
+			repairSummary.NumFailed++
+			continue
+		}
+		repairSummary.NumRepaired++
+	}
+
+	// track count for pending replicas
+	for blkID, hs := range pendingBlockReplicas {
+		if hs.empty() {
+			continue
+		}
+		for _, hbm := range *hs {
+			missingErr := fmt.Errorf(
+				"replica requested but not received [ host = %v, id = %v, block_start = %v ]",
+				hbm.Host.String(), hashToIDMap[blkID.idHash], blkID.start)
+			multiErr = multiErr.Add(missingErr)
+			logger.Warnf("%v", missingErr)
+			repairSummary.NumPending++
+		}
+	}
+
+	// track error if blocksIter is unhappy
+	if err := blocksIter.Err(); err != nil {
+		multiErr.Add(err)
+		logger.Warnf("repair iterator terminated with error: %v", err)
+	}
+
+	// TODO(prateek): better handling of repaired shard writes
+	// 	- write a new version of the file for the timestamp, keep last 'n' versions, do NOT delete old version before writing a new version
+	// 	- consider not always writing, factor in a minimum number of blocks repaired, and number of blocks still requiring repair
+
+	if err := multiErr.FinalError(); err == nil {
+		return repairSummary, nil
+	}
+	return repairSummary, multiErr
```

Comment:
You can probably just `return repairSummary, multiErr.FinalError()` here and avoid the nil check as it will have the same result